lib/code_writer.rb

```diff
@@ -4,33 +4,28 @@ def initialize(out)
     @label_counter = 0
   end
 
-  SEGMENT_TO_SYMBOL_HASH = {
+  DYNAMIC_SEGMENT_BASE_ADDRESS = {
```

Comment:
The name `DYNAMIC_SEGMENT_BASE_ADDRESS` is a big improvement and I appreciate the consistency with `STATIC_SEGMENT_BASE_ADDRESS`, so I’m happy to stick with it if you are.

But I’d like to highlight a small distinction to make sure you’re comfortable with it: `ARG` (for example) is not the base address of the `argument` segment, it’s the _address of the memory location_ where that base address is stored. This is a meaningfully different situation from `STATIC_SEGMENT_BASE_ADDRESS`, where e.g. `5` really is the base address of the `temp` segment.

You already had to understand that to get the implementation working in the first place, but it’s the sort of fiddly detail you could easily forget later, and the naming of these variables doesn’t give your future self any clue that a subtle distinction exists. So an alternative might be something like `DYNAMIC_SEGMENT_POINTER_ADDRESS` — I’ve chosen the word “[pointer](https://en.wikipedia.org/wiki/Pointer_(computer_programming))” out of nowhere and perhaps it isn’t immediately obvious what it means in this context, but it’s _different from_ the word “[base](https://en.wikipedia.org/wiki/Base_address)” in the other variable name so it at least provides a hint that there’s something different about this hash even if it doesn’t constitute a full explanation.